server/comms/rpc/rpctypes.go

```diff
@@ -0,0 +1,381 @@
+// provided in case the caller wants to construct raw responses for some reason.
+// Typically callers will instead want to create the fully marshalled JSON-RPC
+// response to send over the wire with the MarshalResponse function.
+func NewResponse(id interface{}, result interface{}, rpcErr *RPCError) (*Response, error) {
+	if !IsValidIDType(id) {
+		return nil, fmt.Errorf("the id of type '%T' is invalid", id)
+	}
+
+	marshalledResult, err := json.Marshal(result)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Response{
+		Jsonrpc: JSONRPCVersion,
+		Result:  marshalledResult,
+		Error:   rpcErr,
+		ID:      &id,
+	}, nil
+}
+
+// MatchNotification is the params for a DEX-originating MatchMethod request.
+type MatchNotification struct {
+	signable
+	OrderID  string `json:"orderid"`
+	MatchID  string `json:"matchid"`
+	Quantity uint64 `json:"quantity"`
+	Rate     uint64 `json:"rate"`
+	Address  string `json:"address"`
+	Time     uint64 `json:"timestamp"`
+}
+
+// Check that MatchNotification satisfies Signable interface.
+var _ Signable = (*MatchNotification)(nil)
+
+// Serialize serializes the MatchNotification data.
+func (params *MatchNotification) Serialize() ([]byte, error) {
+	// Match serialization is orderid (32) + matchid (32) + quantity (8) + rate (8)
+	// + address (variable, guess 35). Sum = 115
+	s := make([]byte, 0, 91)
+
+	// OrderID
+	oid, err := hex.DecodeString(params.OrderID)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding hex '%s': %v", params.OrderID, err)
+	}
+	s = append(s, oid...)
+
+	// Everything else.
+	matchID, err := hex.DecodeString(params.MatchID)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding match ID %s: %v", params.MatchID, err)
+	}
+	s = append(s, matchID...)
+	s = append(s, uint64Bytes(params.Quantity)...)
+	s = append(s, uint64Bytes(params.Rate)...)
+	s = append(s, []byte(params.Address)...)
+	return s, nil
+}
+
+// InitParams are the params for a client-originating InitMethod request.
+type InitParams struct {
+	signable
+	OrderID  string `json:"orderid"`
+	MatchID  string `json:"matchid"`
+	TxID     string `json:"txid"`
+	Vout     uint32 `json:"vout"`
+	Time     uint64 `json:"timestamp"`
+	Contract string `json:"contract"`
+}
+
+var _ Signable = (*InitParams)(nil)
+
+// Serialize serializes the InitParams data.
+func (params *InitParams) Serialize() ([]byte, error) {
+	// Init serialization is orderid (32) + matchid (32) + txid (probably 64) +
+	// vout (4) + timestamp (8) + contract (97 ish). Sum = 205
+	s := make([]byte, 0, 237)
+
+	oid, err := hex.DecodeString(params.OrderID)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding order id '%s': %v", params.OrderID, err)
+	}
+	s = append(s, oid...)
+	matchID, err := hex.DecodeString(params.MatchID)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding match ID %s: %v", params.MatchID, err)
+	}
+	s = append(s, matchID...)
+	txid := []byte(params.TxID)
```

Comment:
The spec has this listed as the "transaction hash", but I think it makes more sense to use the utf-8 encoding of the txid because the DEX deals with this as a string, and since the tx hash is not simply the bytes represented by the string, at least for Decred and Bitcoin.

Reply:
Two alternative here

1. If it's safe to assume that the string will always be hex, we could just use the bytes as-is, without concern for whether they are reversed by backend. The effect would be that most hashes are reversed in the signature structures from what they would be in common usage.
2. We could add a method to `DEXAsset` to convert a string txid to a hash **without requesting the transaction**. Right now, you can't do that.

Reply:
I commented in my initial review, but now I see the issue here is that take asset's transaction hash might be error prone because of the reversal (or not) of the bytes order?